src/Common/src/Interop/Unix/libcurl/Interop.SafeCurlHandle.cs

```diff
@@ -56,11 +96,175 @@ public static void DisposeAndClearHandle(ref SafeCurlMultiHandle curlHandle)
+
+                    // Always include special fd in the poll set
+                    pollFds[0].fd = _specialFds[libc.ReadEndOfPipe];
+                    pollFds[0].events = PollFlags.POLLIN | PollFlags.POLLOUT;
+                    int i = 1;
+                    foreach (int fd in _fdSet)
+                    {
+                        pollFds[i].fd = fd;
+                        pollFds[i].events = PollFlags.POLLIN | PollFlags.POLLOUT;
+                        i++;
+                    }
+                }
+
+                unsafe
+                {
+                    fixed (libc.pollfd* fds = pollFds)
+                    {
+                        int numFds = libc.poll(fds, (uint)count, -1);
+                        lock (this)
+                        {
+                            if (0 == _fdCount)
+                            {
+                                return;
+                            }
+                        }
+                        if (numFds <= 0)
+                        {
+                            Debug.Assert(numFds != 0); // Since timeout is infinite
+                            // ignore errors
+                            return;
+                        }
+
+                        for (int i = 0, j = 0; i < count && j < numFds; i++)
+                        {
+                            if (fds[i].revents == 0)
+                            {
+                                continue;
+                            }
+                            if (i == 0)
+                            {
+                                if (ReadSpecialFd(fds[i].revents) < 0)
+                                {
+                                    // TODO: How to handle errors?
+                                    throw new InvalidOperationException("Cannot read data: " + Marshal.GetLastWin32Error());
+                                }
+                                numFds--;
+                            }
+                            else
+                            {
+                                libc.pollfd polledFd = new libc.pollfd
+                                {
+                                    fd = fds[i].fd,
+                                    events = fds[i].events,
+                                    revents = fds[i].revents,
+                                };
+                                readyFds.Add(polledFd);
+                                j++;
+                            }
+                        }
+                    }
+                }
+            }
+
+            internal void SignalFdSetChange(int fd, bool isRemove)
+            {
+                Debug.Assert(Monitor.IsEntered(this));
+                if (isRemove)
+                {
+                    if (!_fdSet.Remove(fd))
+                    {
+                        return;
+                    }
+                }
+                else
+                {
+                    _fdSet.Add(fd);
```

Comment:
What if it's already in the set?  Should this if/else be more like:
```C#
bool changed = isRemove ?
    _fdSet.Remove(fd) :
    _fdSet.Add(fd);
if (!changed)
{
    return;
}
```
or if we never expect Add to return false, should we instead Debug.Assert that its result is true?